earth_enterprise/src/common/khxml/khxml.cpp

```diff
@@ -101,66 +125,195 @@ void setDefaultValues()
+public:
+    static purgeGuard& instance()
+    {
+        static purgeGuard _instance;
+        return _instance;
+    }
+
+    static void increment()
+    {
+        khLockGuard guard(pLock);
+        ++count;
+    }
+
+    static void decrement()
+    {
+        khLockGuard guard(pLock);
+        --count;
+    }
+
+    static uint32_t getCount()
+    {
+        khLockGuard guard(pLock);
+        return count;
+    }
+
+    static void lock()
+    {
+        khLockGuard guard(pLock);
+        _lock = true;
+    }
+
+    static void unlock()
+    {
+        khLockGuard guard(pLock);
+        _lock = false;
+    }
+
+    static bool isLocked()
+    {
+        khLockGuard guard(pLock);
+        return _lock;
+    }
+};
+bool purgeGuard::_lock = false;
+khMutexBase purgeGuard::pLock = KH_MUTEX_BASE_INITIALIZER;
+uint32_t purgeGuard::count = 0;
+
+void ReInitializeXerces()
+{
+    khLockGuard guard(reinitLock);
+    try
+    {
+        XMLPlatformUtils::Terminate();
+    }
+    catch(...)
+    {
+        notify(NFY_WARN, "Failure to terminate in ReInitializeXerces()");
+    }
+    initXercesValues();
+}
+
+
 // This is used only in the following function
 class UsingXMLGuard
 {
   friend void InitializeXMLLibrary() throw();
 
   UsingXMLGuard(void) throw() {
-    setDefaultValues();
-    std::string fn(XMLConfigFile);
-    try {
-      for (const auto& a : options)
-      {
-          config_parser.addOption(a);
-      }
-      config_parser.parse(fn);
-      config_parser.validateIntegerValues();
-      for (const auto& a : config_parser)
-      {
-          // will only set the values that are present, otherwise will stay at defaults
-          if (a.first == INIT_HEAP_SIZE)
-              initialDOMHeapAllocSize = std::stol(a.second);
-          else if (a.first == MAX_HEAP_SIZE)
-              maxDOMHeapAllocSize = std::stol(a.second);
-          else if (a.first == BLOCK_SIZE)
-              maxDOMSubAllocationSize = std::stol(a.second);
-      }
-    } catch (const khConfigFileParserException& e) {
-      notify(NFY_DEBUG, "%s , using default xerces init values", e.what());
-    }
-    
-    try {
-      try {
-         validateXMLParameters();
-      } catch (const XmlParamsException& e) {
-         notify(NFY_DEBUG, "%s, using default xerces init values", e.what());
-         setDefaultValues();
-      }
-      XMLPlatformUtils::Initialize(initialDOMHeapAllocSize,
-                                   maxDOMHeapAllocSize,
-                                   maxDOMSubAllocationSize);
-      notify(NFY_DEBUG, "XML initialization values: %s=%zu %s=%zu %s=%zu",
-             "initialDOMHeapAllocSize", initialDOMHeapAllocSize,
-             "maxDOMHeapAllocSize", maxDOMHeapAllocSize,
-             "maxDOMSubAllocationSize", maxDOMSubAllocationSize);
-    } catch(const XMLException& toCatch) {
-      notify(NFY_FATAL, "Unable to initialize Xerces: %s",
-             FromXMLStr(toCatch.getMessage()).c_str());
-    }
+    initXercesValues();
```

Comment:
For complete safety, this and the terminate call in the destructor should be called while the reinitLock is held so that we're absolutely certain multiple Initializes and Terminates can't happen at the same time.